crates/subspace-farmer/src/farming.rs

```diff
@@ -0,0 +1,226 @@
+use crate::commitments::Commitments;
+use crate::common::Salt;
+use crate::identity::Identity;
+use crate::plot::Plot;
+use crate::rpc::{ProposedProofOfReplicationResponse, RpcClient, SlotInfo, Solution};
+use anyhow::{anyhow, Result};
+use futures::{future, future::Either};
+use log::{debug, error, info, trace};
+use std::time::Instant;
+use subspace_core_primitives::crypto;
+
+/// Farming Instance to store the necessary information for the farming operations,
+/// and also a channel to stop/pause the background farming task
+pub struct Farming {
+    plot: Plot,
+    commitments: Commitments,
+    client: RpcClient,
+    identity: Identity,
+    sender: Option<async_oneshot::Sender<()>>,
+}
+
+#[derive(Default)]
+struct Salts {
+    current: Option<Salt>,
+    next: Option<Salt>,
+}
+
+impl Farming {
+    /// returns an instance of farming, and also starts a concurrent background farming task
+    pub fn start(
+        plot: Plot,
+        commitments: Commitments,
+        client: RpcClient,
+        identity: Identity,
+    ) -> Self {
+        let (sender, receiver) = async_oneshot::oneshot();
+
+        let farming = Farming {
+            plot,
+            commitments,
+            client,
+            identity,
+            sender: Some(sender),
+        };
+
+        tokio::spawn(background_farming(
+            farming.client.clone(),
+            farming.plot.clone(),
+            farming.commitments.clone(),
+            farming.identity.clone(),
+            receiver,
+        ));
+
+        farming
+    }
+}
+
+async fn background_farming(
+    client: RpcClient,
+    plot: Plot,
+    commitments: Commitments,
+    identity: Identity,
+    receiver: async_oneshot::Receiver<()>,
+) -> Result<()> {
```

Comment:
Do you really need `Result<>` here? It is not consumed by anything anyway, so probably doesn't matter?